<a href="https://colab.research.google.com/github/gcanapa/base_mnist_usando_rede_neural_backpropagation_e_cnn/blob/main/base_mnist_usando_rede_neural_backpropagation_e_cnn.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Importando as bibliotecas necessárias

In [1]:
import numpy as np
import tensorflow as tf
from tensorflow.keras.datasets import mnist

Carregando dados do MNIST


In [2]:
(x_train, y_train), (x_test, y_test) = mnist.load_data()

11490434/11490434 [==============================] - 0s 0us/step


Pré-processando dados 

In [3]:
# Normalizar os pixels para o intervalo [0, 1]
x_train = x_train / 255.0
x_test = x_test / 255.0

# Redimensionar as imagens para o tamanho desejado (por exemplo, 28x28 pixels)
x_train = np.expand_dims(x_train, axis=-1)
x_test = np.expand_dims(x_test, axis=-1)

# Codificar os rótulos usando one-hot encoding
y_train = tf.keras.utils.to_categorical(y_train, num_classes=10)
y_test = tf.keras.utils.to_categorical(y_test, num_classes=10)


**Defindo a arquitetura da Backpropagation**
# **Teste 1:**



In [ ]:
model = tf.keras.models.Sequential([
    tf.keras.layers.Flatten(input_shape=(28, 28, 1)),
    tf.keras.layers.Dense(128, activation='relu'),
    tf.keras.layers.Dense(10, activation='softmax')
])

Nesse exemplo, a rede neural possui uma camada de entrada flattenizada (transformando a imagem 28x28 em um vetor de tamanho 784), uma camada oculta com 128 neurônios usando a função de ativação ReLU, e uma camada de saída com 10 neurônios usando a função de ativação softmax para classificação multiclasse.

Compilando o modelo

In [ ]:
model.compile(optimizer='sgd', loss='categorical_crossentropy', metrics=['accuracy'])


Utilizando o otimizador SGD (Stochastic Gradient Descent) e a função de perda categorical_crossentropy, que é comumente usada em problemas de classificação multiclasse.

Treinando o modelo usando a Backpropagation

In [ ]:
model.fit(x_train, y_train, batch_size=32, epochs=10)

Epoch 1/10
1875/1875 [==============================] - 7s 3ms/step - loss: 0.6271 - accuracy: 0.8432
Epoch 2/10
1875/1875 [==============================] - 7s 4ms/step - loss: 0.3326 - accuracy: 0.9082
Epoch 3/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.2865 - accuracy: 0.9199
Epoch 4/10
1875/1875 [==============================] - 7s 4ms/step - loss: 0.2566 - accuracy: 0.9283
Epoch 5/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.2339 - accuracy: 0.9348
Epoch 6/10
1875/1875 [==============================] - 7s 4ms/step - loss: 0.2153 - accuracy: 0.9403
Epoch 7/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.1996 - accuracy: 0.9446
Epoch 8/10
1875/1875 [==============================] - 7s 4ms/step - loss: 0.1863 - accuracy: 0.9481
Epoch 9/10
1875/1875 [==============================] - 5s 3ms/step - loss: 0.1745 - accuracy: 0.9511
Epoch 10/10
1875/1875 [==============================] - 7s 4ms/step - loss: 0.164

Estamos treinando o modelo com os dados de treinamento (x_train e y_train) durante 10 épocas, com um tamanho de lote (batch_size) de 32.

Durante o treinamento, o TensorFlow irá automaticamente realizar a propagação para frente (forward propagation), onde os dados são passados pela rede neural, e a retropropagação do erro (backward propagation), onde os gradientes são calculados e os pesos da rede são atualizados usando o algoritmo de otimização escolhido (SGD neste caso).

Avaliando o desempenho:

In [ ]:
test_loss, test_accuracy = model.evaluate(x_test, y_test)


313/313 [==============================] - 1s 3ms/step - loss: 0.1602 - accuracy: 0.9552


A "loss" (perda) é uma medida do quão bem o modelo está se ajustando aos dados de treinamento. Quanto menor o valor da perda, melhor o modelo está se ajustando aos dados. O valor de 0.1602 indica uma boa adequação do modelo aos dados de treinamento.

A "accuracy" (precisão) é uma métrica que mede a taxa de acertos do modelo em relação aos rótulos verdadeiros dos exemplos de treinamento. Ela é expressa como uma porcentagem, variando de 0 a 1 (ou 0% a 100%). Uma precisão de 0.9552 ou 95.52% indica que o modelo está prevendo corretamente a classe correta para cerca de 95.52% dos exemplos de treinamento



---

# **Teste 2:**

Substituindo a função de ativação ReLU por tangente hiperbólica (tanh) na camada oculta.

In [ ]:
model = tf.keras.models.Sequential([
    tf.keras.layers.Flatten(input_shape=(28, 28, 1)),
    tf.keras.layers.Dense(128, activation='tanh'),
    tf.keras.layers.Dense(10, activation='softmax')
])

Otimizador e função de perda:

In [ ]:
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

Utilizando o otimizador Adam e a função de perda sparse_categorical_crossentropy, que é usada quando os rótulos não estão codificados em one-hot encoding.

A função sparse_categorical_crossentropy espera rótulos na forma de inteiros, em vez de codificação one-hot. Portanto, você precisa garantir que os rótulos y_train estejam no formato correto antes de treinar o modelo.

Uma maneira de corrigir isso é ajustar a forma dos rótulos usando a função np.argmax() antes de chamar model.fit(). 

In [ ]:
y_train = np.argmax(y_train, axis=1)

 Utilizando um tamanho de lote de 64 e treinando o modelo por 20 épocas.

In [ ]:
model.fit(x_train, y_train, batch_size=64, epochs=20)

Epoch 1/20
938/938 [==============================] - 6s 6ms/step - loss: 0.3278 - accuracy: 0.1001
Epoch 2/20
938/938 [==============================] - 4s 4ms/step - loss: 0.1661 - accuracy: 0.0992
Epoch 3/20
938/938 [==============================] - 5s 5ms/step - loss: 0.1185 - accuracy: 0.0992
Epoch 4/20
938/938 [==============================] - 5s 6ms/step - loss: 0.0906 - accuracy: 0.0987
Epoch 5/20
938/938 [==============================] - 4s 5ms/step - loss: 0.0712 - accuracy: 0.0989
Epoch 6/20
938/938 [==============================] - 5s 6ms/step - loss: 0.0573 - accuracy: 0.0990
Epoch 7/20
938/938 [==============================] - 4s 5ms/step - loss: 0.0458 - accuracy: 0.0989
Epoch 8/20
938/938 [==============================] - 4s 5ms/step - loss: 0.0376 - accuracy: 0.0991
Epoch 9/20
938/938 [==============================] - 6s 6ms/step - loss: 0.0298 - accuracy: 0.0990
Epoch 10/20
938/938 [==============================] - 4s 5ms/step - loss: 0.0244 - accuracy: 0.0989

Convertendo os dados para o formato de inteiros

In [ ]:
y_test = np.argmax(y_test, axis=1)

Avaliando o desempenho

In [ ]:
test_loss, test_accuracy = model.evaluate(x_test, y_test)

313/313 [==============================] - 1s 2ms/step - loss: 0.0836 - accuracy: 0.0990


Loss: O valor de 0.0836 indica a média da perda para os exemplos do conjunto de teste. Essa perda é calculada usando a função de perda que você especificou ao compilar o modelo. Quanto menor o valor da perda, melhor o modelo está performando nos dados de teste.

Accuracy: O valor de 0.0990 indica a precisão média do modelo para os exemplos do conjunto de teste. Essa precisão é calculada como a taxa de acertos do modelo em relação aos rótulos verdadeiros dos exemplos do conjunto de teste. Uma precisão de 0.0990 ou 9.90% indica que o modelo está prevendo corretamente a classe correta apenas para cerca de 9.90% dos exemplos do conjunto de teste.



---


# **Teste 3:**

Aumentando a quantidade de neuronios na camada oculta para 256

In [4]:
model = tf.keras.models.Sequential([
    tf.keras.layers.Flatten(input_shape=(28, 28, 1)),
    tf.keras.layers.Dense(256, activation='relu'),
    tf.keras.layers.Dense(10, activation='softmax')
])

In [5]:
model.compile(optimizer='sgd', loss='categorical_crossentropy', metrics=['accuracy'])

Vamos utilizar um batch_size de 64 e 20 épocas

In [6]:
model.fit(x_train, y_train, batch_size=64, epochs=20)

Epoch 1/20
938/938 [==============================] - 6s 5ms/step - loss: 0.8289 - accuracy: 0.8042
Epoch 2/20
938/938 [==============================] - 4s 4ms/step - loss: 0.4057 - accuracy: 0.8907
Epoch 3/20
938/938 [==============================] - 4s 4ms/step - loss: 0.3436 - accuracy: 0.9049
Epoch 4/20
938/938 [==============================] - 5s 5ms/step - loss: 0.3109 - accuracy: 0.9131
Epoch 5/20
938/938 [==============================] - 4s 4ms/step - loss: 0.2882 - accuracy: 0.9191
Epoch 6/20
938/938 [==============================] - 4s 4ms/step - loss: 0.2702 - accuracy: 0.9236
Epoch 7/20
938/938 [==============================] - 5s 5ms/step - loss: 0.2551 - accuracy: 0.9284
Epoch 8/20
938/938 [==============================] - 4s 4ms/step - loss: 0.2420 - accuracy: 0.9327
Epoch 9/20
938/938 [==============================] - 4s 4ms/step - loss: 0.2307 - accuracy: 0.9352
Epoch 10/20
938/938 [==============================] - 4s 5ms/step - loss: 0.2203 - accuracy: 0.9382

In [7]:
test_loss, test_accuracy = model.evaluate(x_test, y_test)

313/313 [==============================] - 1s 2ms/step - loss: 0.1519 - accuracy: 0.9552


Loss: O valor de 0.1519 indica uma boa adequação do modelo aos dados de treinamento.

Accuracy: Uma precisão de 0.9552 ou 95.52% indica que o modelo está prevendo corretamente a classe correta para cerca de 95.52% dos exemplos de treinamento



---

# Utilizando rede neural CNN

#Teste 1

Essas linhas de código abaixo importam as bibliotecas necessárias. O mnist é importado da biblioteca Keras.datasets e é usado para carregar o conjunto de dados MNIST. O Sequential é importado da biblioteca Keras.models e é usado para criar uma sequência de camadas da rede neural. Conv2D, MaxPooling2D, Flatten e Dense são importados da biblioteca Keras.layers e são usados para definir diferentes tipos de camadas da rede neural. to_categorical é importado da biblioteca Keras.utils e é usado para converter os rótulos de classe em formato one-hot.

In [8]:
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Conv2D, MaxPooling2D, Flatten, Dense
from keras.utils import to_categorical

Essa linha de código abaixo carregam o conjunto de dados MNIST usando a função load_data() do módulo mnist do Keras.datasets. Ele retorna duas tuplas: uma contendo as imagens de treinamento (x_train) e os rótulos correspondentes (y_train), e a outra contendo as imagens de teste (x_test) e os rótulos correspondentes (y_test).

In [9]:
(x_train, y_train), (x_test, y_test) = mnist.load_data()

Essas linhas de código abaixo redimensionam e normalizam as imagens de treinamento e teste. As imagens originais têm uma forma de (número de amostras, 28, 28), mas a camada de convolução espera uma entrada com uma dimensão adicional para representar os canais de cor. Portanto, usamos reshape para adicionar uma dimensão adicional de 1, resultando em uma forma de (número de amostras, 28, 28, 1). Em seguida, as imagens são convertidas para o tipo de dados float32 e normalizadas dividindo cada pixel por 255.0, para que os valores estejam na faixa entre 0 e 1.

In [10]:
x_train = x_train.reshape((-1, 28, 28, 1)).astype('float32') / 255.0
x_test = x_test.reshape((-1, 28, 28, 1)).astype('float32') / 255.0

Essas linhas de código abaixo convertem os rótulos de classe em formato one-hot usando a função to_categorical do Keras.utils. O formato one-hot codifica as classes como vetores binários, onde cada posição do vetor representa uma classe possível e é marcada com 1 se a amostra pertence àquela classe e 0 caso contrário.

In [11]:
y_train = to_categorical(y_train)
y_test = to_categorical(y_test)

Essas linhas de código abaixo definem a arquitetura da rede neural. A classe Sequential é inicializada para criar um modelo sequencial. Em seguida, diferentes camadas são adicionadas ao modelo usando o método add(). Primeiro, uma camada convolucional com 32 filtros de tamanho 3x3 e função de ativação ReLU é adicionada, com uma forma de entrada (28, 28, 1). Em seguida, uma camada de max pooling com janela 2x2 é adicionada para reduzir a dimensionalidade. Essa sequência de uma camada convolucional seguida por uma camada de max pooling é repetida com 64 filtros na segunda camada convolucional. Depois disso, a saída da camada de convolução é achatada para uma representação vetorial única usando a camada Flatten(). Duas camadas totalmente conectadas são adicionadas: uma camada oculta com 64 unidades e função de ativação ReLU, e uma camada de saída com 10 unidades (correspondendo às 10 classes do conjunto de dados MNIST) e função de ativação softmax.



In [12]:
model = Sequential()
model.add(Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)))
model.add(MaxPooling2D((2, 2)))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(MaxPooling2D((2, 2)))
model.add(Flatten())
model.add(Dense(64, activation='relu'))
model.add(Dense(10, activation='softmax'))

Essa linha de código abaixo compilam o modelo. O otimizador 'adam' é usado, que é uma variante do otimizador de descida do gradiente estocástico (SGD). A função de perda 'categorical_crossentropy' é usada, pois temos várias classes categóricas. Além disso, a métrica de avaliação 'accuracy' é especificada para monitorar o desempenho durante o treinamento.



In [13]:
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

Essa linha de código abaixo treinam o modelo. A função fit() é usada para ajustar o modelo aos dados de treinamento. x_train e y_train são os dados de treinamento, epochs=10 define o número de épocas de treinamento, batch_size=128 especifica o tamanho do lote usado durante o treinamento e validation_data=(x_test, y_test) indica os dados de validação usados durante o treinamento para avaliar o desempenho do modelo.






In [14]:
model.fit(x_train, y_train, epochs=10, batch_size=128, validation_data=(x_test, y_test))

Epoch 1/10
469/469 [==============================] - 43s 91ms/step - loss: 0.2373 - accuracy: 0.9342 - val_loss: 0.0608 - val_accuracy: 0.9808
Epoch 2/10
469/469 [==============================] - 43s 92ms/step - loss: 0.0612 - accuracy: 0.9816 - val_loss: 0.0581 - val_accuracy: 0.9818
Epoch 3/10
469/469 [==============================] - 44s 95ms/step - loss: 0.0435 - accuracy: 0.9867 - val_loss: 0.0407 - val_accuracy: 0.9871
Epoch 4/10
469/469 [==============================] - 43s 92ms/step - loss: 0.0340 - accuracy: 0.9892 - val_loss: 0.0360 - val_accuracy: 0.9880
Epoch 5/10
469/469 [==============================] - 42s 89ms/step - loss: 0.0275 - accuracy: 0.9913 - val_loss: 0.0289 - val_accuracy: 0.9912
Epoch 6/10
469/469 [==============================] - 43s 92ms/step - loss: 0.0225 - accuracy: 0.9929 - val_loss: 0.0292 - val_accuracy: 0.9906
Epoch 7/10
469/469 [==============================] - 42s 89ms/step - loss: 0.0184 - accuracy: 0.9939 - val_loss: 0.0375 - val_accuracy:

Para obter o resultado de desempenho da rede neural treinada, você pode usar o método evaluate() do modelo. O método evaluate() calcula a perda e a métrica especificada durante o treinamento (nesse caso, a precisão) nos dados de teste fornecidos (x_test e y_test). O valor da perda e a precisão são então retornados.

In [15]:
loss, accuracy = model.evaluate(x_test, y_test)
print('Test Loss:', loss)
print('Test Accuracy:', accuracy)

313/313 [==============================] - 2s 7ms/step - loss: 0.0382 - accuracy: 0.9891
Test Loss: 0.03818608820438385
Test Accuracy: 0.9890999794006348


Esses números indicam um bom desempenho do modelo treinado. A perda é baixa e a precisão é alta, sugerindo que a rede neural está fazendo previsões precisas para a maioria das amostras de teste.

loss: 0.0382 indica o valor da perda obtido no conjunto de teste. A perda é de aproximadamente 0.0382.

accuracy: 0.9891 indica a precisão alcançada no conjunto de teste. A precisão é de aproximadamente 0.9891, o que significa que cerca de 98,91% das amostras de teste foram classificadas corretamente pela rede neural.



---
#Teste 2



Neste exemplo, estamos adaptando a arquitetura da CNN para lidar com imagens de tamanho 28x28 pixels e uma única camada de cor (escala de cinza). Portanto, fazemos algumas modificações, como o ajuste da forma de entrada, que agora é (28, 28, 1), e a remoção da camada de pooling adicional para preservar mais informações. Alteramos também a função de ativação para sigmoid.

In [18]:
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Conv2D, MaxPooling2D, Flatten, Dense
from keras.utils import to_categorical

# Carregando o conjunto de dados MNIST
(x_train, y_train), (x_test, y_test) = mnist.load_data()

# Pré-processamento dos dados
x_train = x_train.reshape(-1, 28, 28, 1).astype('float32') / 255.0
x_test = x_test.reshape(-1, 28, 28, 1).astype('float32') / 255.0
y_train = to_categorical(y_train)
y_test = to_categorical(y_test)

# Criando o modelo da CNN
model = Sequential()
model.add(Conv2D(32, (3, 3), activation='sigmoid', input_shape=(28, 28, 1)))
model.add(MaxPooling2D((2, 2)))
model.add(Conv2D(64, (3, 3), activation='sigmoid'))
model.add(MaxPooling2D((2, 2)))
model.add(Flatten())
model.add(Dense(64, activation='sigmoid'))
model.add(Dense(10, activation='softmax'))

# Compilando e treinando o modelo
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
model.fit(x_train, y_train, epochs=10, batch_size=128, validation_data=(x_test, y_test))

# Avaliando o desempenho do modelo nos dados de teste
print('\n')
loss, accuracy = model.evaluate(x_test, y_test)
print('\nTest Loss:', loss)
print('Test Accuracy:', accuracy)


Epoch 1/10
469/469 [==============================] - 46s 97ms/step - loss: 1.2412 - accuracy: 0.5919 - val_loss: 0.3716 - val_accuracy: 0.9020
Epoch 2/10
469/469 [==============================] - 43s 91ms/step - loss: 0.2747 - accuracy: 0.9248 - val_loss: 0.1787 - val_accuracy: 0.9514
Epoch 3/10
469/469 [==============================] - 43s 92ms/step - loss: 0.1567 - accuracy: 0.9564 - val_loss: 0.1233 - val_accuracy: 0.9639
Epoch 4/10
469/469 [==============================] - 44s 93ms/step - loss: 0.1113 - accuracy: 0.9683 - val_loss: 0.0923 - val_accuracy: 0.9737
Epoch 5/10
469/469 [==============================] - 43s 92ms/step - loss: 0.0875 - accuracy: 0.9752 - val_loss: 0.0798 - val_accuracy: 0.9752
Epoch 6/10
469/469 [==============================] - 45s 97ms/step - loss: 0.0734 - accuracy: 0.9784 - val_loss: 0.0663 - val_accuracy: 0.9798
Epoch 7/10
469/469 [==============================] - 46s 97ms/step - loss: 0.0634 - accuracy: 0.9809 - val_loss: 0.0619 - val_accuracy:

Esses números indicam um bom desempenho do modelo treinado. A perda é baixa e a precisão é alta, sugerindo que a rede neural está fazendo previsões precisas para a maioria das amostras de teste.

loss: 0.0435 indica o valor da perda obtido no conjunto de teste. A perda é de aproximadamente 0.0435.

accuracy: 0.9860 indica a precisão alcançada no conjunto de teste. A precisão é de aproximadamente 0.9860, o que significa que cerca de 98,60% das amostras de teste foram classificadas corretamente pela rede neural.



---
#Teste 3


Neste exemplo, há algumas alterações nos parâmetros da arquitetura da CNN em comparação com o exemplo anterior:

A primeira camada convolucional agora usa 16 filtros com tamanho 5x5.
A segunda camada convolucional agora usa 32 filtros com tamanho 3x3.

In [17]:
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Conv2D, MaxPooling2D, Flatten, Dense
from keras.utils import to_categorical

# Carregando o conjunto de dados MNIST
(x_train, y_train), (x_test, y_test) = mnist.load_data()

# Pré-processamento dos dados
x_train = x_train.reshape(-1, 28, 28, 1).astype('float32') / 255.0
x_test = x_test.reshape(-1, 28, 28, 1).astype('float32') / 255.0
y_train = to_categorical(y_train)
y_test = to_categorical(y_test)

# Criando o modelo da CNN com uma variação de parâmetros
model = Sequential()
model.add(Conv2D(16, (5, 5), activation='relu', input_shape=(28, 28, 1)))
model.add(MaxPooling2D((2, 2)))
model.add(Conv2D(32, (3, 3), activation='relu'))
model.add(MaxPooling2D((2, 2)))
model.add(Flatten())
model.add(Dense(64, activation='relu'))
model.add(Dense(10, activation='softmax'))

# Compilando e treinando o modelo
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
model.fit(x_train, y_train, epochs=10, batch_size=128, validation_data=(x_test, y_test))

# Avaliando o desempenho do modelo nos dados de teste
print('\n')
loss, accuracy = model.evaluate(x_test, y_test)
print('\nTest Loss:', loss)
print('Test Accuracy:', accuracy)


Epoch 1/10
469/469 [==============================] - 27s 56ms/step - loss: 0.2701 - accuracy: 0.9218 - val_loss: 0.0915 - val_accuracy: 0.9713
Epoch 2/10
469/469 [==============================] - 25s 54ms/step - loss: 0.0715 - accuracy: 0.9785 - val_loss: 0.0483 - val_accuracy: 0.9851
Epoch 3/10
469/469 [==============================] - 28s 60ms/step - loss: 0.0510 - accuracy: 0.9847 - val_loss: 0.0434 - val_accuracy: 0.9860
Epoch 4/10
469/469 [==============================] - 26s 55ms/step - loss: 0.0419 - accuracy: 0.9874 - val_loss: 0.0435 - val_accuracy: 0.9861
Epoch 5/10
469/469 [==============================] - 25s 54ms/step - loss: 0.0354 - accuracy: 0.9887 - val_loss: 0.0326 - val_accuracy: 0.9910
Epoch 6/10
469/469 [==============================] - 25s 54ms/step - loss: 0.0294 - accuracy: 0.9908 - val_loss: 0.0335 - val_accuracy: 0.9892
Epoch 7/10
469/469 [==============================] - 25s 54ms/step - loss: 0.0257 - accuracy: 0.9918 - val_loss: 0.0481 - val_accuracy:

loss: 0.0288 indica o valor da perda obtido no conjunto de teste. A perda é de aproximadamente 0.0288.

accuracy: 0.9907 indica a precisão alcançada no conjunto de teste. A precisão é de aproximadamente 0.9907, o que significa que cerca de 99,07% das amostras de teste foram classificadas corretamente pela rede neural.